force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,201 @@
+        List<DonationHistoryController.DonationHistory> partialAndSoftCredit = verifyPartialAndSoftCreditCondition(softCredit, partialSoftCredit);
+        
+        donationHistory.addAll(hardCredit);
+        donationHistory.addAll(partialAndSoftCredit);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyHardCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory();
+                hardCredit.id = ocr.Id;
+                hardCredit.name = ocr.Opportunity.Name;
+                hardCredit.isPrimary = ocr.isPrimary;
+                hardCredit.closeDate = ocr.Opportunity.CloseDate;
+                hardCredit.role = ocr.Role;
+                hardCredit.amount = ocr.Opportunity.Amount;
+                hardCredit.creditType = HARD_CREDIT;
+                hardCredit.paymentMethod = 'Cash';
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifySoftCreditCondition(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group
+        Filter_Rule__mdt roles = [SELECT Constant__c FROM Filter_Rule__mdt WHERE QualifiedApiName = 'NPSP_OppsSoftCreditRoles'];
+        List<String> softCreditRoles = roles.Constant__c.split(';');
+        List<DonationHistoryController.DonationHistory> softCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (softCreditRoles.contains(ocr.Role) && ocr.isPrimary==false) {
+                DonationHistoryController.DonationHistory softCredit = new DonationHistoryController.DonationHistory();
+                softCredit.id = ocr.Id;
+                softCredit.name = ocr.Opportunity.Name;
+                softCredit.isPrimary = ocr.isPrimary;
+                softCredit.closeDate = ocr.Opportunity.CloseDate;
+                softCredit.role = ocr.Role;
+                softCredit.amount = ocr.Opportunity.Amount;
+                softCredit.creditType = SOFT_CREDIT;
+                softCredit.paymentMethod = 'Cash';
+                softCreditList.add(softCredit);
```

Comment:
This logic is nearly identical the logic on lines 98-107 above. These can be combined into a single method, possibly as the constructor for `DonationHistory`, and then only change the values that require changing (`CreditType`).

Reply:
Constructor was added for the class but the logic that validates the types of credits is different for each case, so it was not modified.